File: mlx-sys/src/bridge/stream.rs
//! Bridge for MLX Device/Stream types and async transforms.
//!
//! Uses cxx shared structs (Device, Stream) for zero-overhead POD value
//! passing. Layout binary-compatible with mlx::core::Device/Stream (same
//! field order, same underlying types). Conversion to/from MLX native types
//! happens in the C++ shim via field-by-field copy.
//!
//! ## DeviceType note
//!
//! cxx shared enums are emitted as `#[repr(transparent)] struct { pub repr: T }`
//! on the Rust side, which does NOT support the `as` primitive cast syntax.
//! To keep the public API ergonomic (`d.device_type as i32`,
//! `DeviceType::Gpu as i32`), `DeviceType` is defined as a native Rust
//! `#[repr(i32)]` enum outside the cxx bridge, and `Device.device_type` uses
//! `i32` as the wire type (layout-identical). The public `ffi` module
//! re-exports `DeviceType` so callers use the canonical `ffi::DeviceType` path.

/// Mirror of `mlx::core::Device::DeviceType`. Native Rust `#[repr(i32)]` enum
/// so that `as i32` primitive casting works. Values match MLX's declaration
/// order in mlx/device.h (cpu=0, gpu=1).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum DeviceType {
    Cpu = 0,
    Gpu = 1,
}

// cxx::bridge generates `unsafe fn` declarations for our pointer-slice variants
// (async_eval_many). The Safety contract is documented in the safe Rust wrapper
// (`mlx::transforms::async_eval`); cxx doesn't propagate doc comments from
// inside the bridge macro.
#[allow(clippy::missing_safety_doc)]
#[cxx::bridge(namespace = "cxx_mlx")]
mod ffi_bridge {
    /// Mirror of `mlx::core::Device`. POD, 8 bytes, layout-compatible with MLX.
    /// `device_type` stored as `i32` (same layout as `DeviceType`'s `#[repr(i32)]`).
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct Device {
        pub device_type: i32,
        pub index: i32,
    }

    /// Mirror of `mlx::core::Stream`. POD, 12 bytes, layout-compatible with MLX.
    /// Streams must be obtained via `default_stream()` / `new_stream()` —
    /// constructing one with arbitrary `index` is undefined behavior in MLX.
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub struct Stream {
        pub index: i32,
        pub device: Device,
    }

    unsafe extern "C++" {
        include!("cxx_mlx_shim/stream.h");

        // Reuse the MlxArray opaque type from the array bridge for async_eval.
        type MlxArray = crate::bridge::array::ffi::MlxArray;

        // === Device queries ===
        fn default_device() -> Device;
        fn set_default_device(d: Device);
        fn is_available(d: Device) -> bool;
        fn device_count(t: i32) -> i32;

        // === Stream lifecycle ===
        fn default_stream(d: Device) -> Stream;
        fn new_stream(d: Device) -> Result<Stream>;
        fn set_default_stream(s: Stream);
        fn get_streams() -> Vec<Stream>;
        fn clear_streams();

        // === Transforms ===
        unsafe fn async_eval_many(arrays: &[*const MlxArray]) -> Result<()>;
        fn synchronize() -> Result<()>;
        fn synchronize_stream(s: Stream) -> Result<()>;
    }
}

/// Public FFI module. Re-exports the cxx bridge items plus `DeviceType`
/// (defined outside the bridge as a native Rust enum to support `as i32` casting).
#[allow(clippy::missing_safety_doc)]
pub mod ffi {
    pub use super::ffi_bridge::{
        async_eval_many, clear_streams, default_device, default_stream, device_count, get_streams,
        is_available, new_stream, set_default_device, set_default_stream, synchronize,
        synchronize_stream, Device, MlxArray, Stream,
    };
    pub use super::DeviceType;
}

File: mlx-sys/shim/src/stream.cc
#include "cxx_mlx_shim/stream.h"

// The cxx-generated header is what defines Device/Stream as concrete types.
// Including it here is what makes the conversion functions have access to
// full type definitions.
#include "mlx-sys/src/bridge/stream.rs.h"

#include "mlx/memory.h"
#include "mlx/transforms.h"

namespace cxx_mlx {

// === Conversions (cxx_mlx ↔ mlx::core) ===
//
// Both representations have identical layout (same field order, same
// underlying integer types). We do field-by-field copy rather than
// reinterpret_cast — explicit and safe, the compiler optimizes it to
// register-level copy.
//
// Note: Device.device_type is stored as int32_t on the Rust side (wire type
// for the bridge), matching MLX's DeviceType underlying type (int/int32_t).

namespace {

mlx::core::Device::DeviceType to_mlx_dtype(int32_t t) {
  // Layout-compatible: the int32_t wire value matches MLX's enum class
  // underlying type (int). Values: Cpu=0, Gpu=1 match mlx/device.h:14-17.
  return static_cast<mlx::core::Device::DeviceType>(t);
}

int32_t from_mlx_dtype(mlx::core::Device::DeviceType t) {
  return static_cast<int32_t>(t);
}

mlx::core::Device to_mlx(Device d) {
  return mlx::core::Device(to_mlx_dtype(d.device_type), d.index);
}

Device from_mlx(const mlx::core::Device& d) {
  return Device{from_mlx_dtype(d.type), d.index};
}

mlx::core::Stream to_mlx(Stream s) {
  return mlx::core::Stream(s.index, to_mlx(s.device));
}

mlx::core::ThreadLocalStream to_mlx_thread_local(Stream s) {
  return mlx::core::ThreadLocalStream(s.index, to_mlx(s.device));
}

Stream from_mlx(const mlx::core::Stream& s) {
  return Stream{s.index, from_mlx(s.device)};
}

}  // namespace

// === Device API ===

Device default_device() {
  return from_mlx(mlx::core::default_device());
}

void set_default_device(Device d) {
  mlx::core::set_default_device(to_mlx(d));
}

bool is_available(Device d) {
  return mlx::core::is_available(to_mlx(d));
}

int32_t device_count(int32_t t) {
  return mlx::core::device_count(to_mlx_dtype(t));
}

// === Stream API ===

Stream default_stream(Device d) {
  return from_mlx(mlx::core::default_stream(to_mlx(d)));
}

Stream new_stream(Device d) {
  return from_mlx(mlx::core::new_stream(to_mlx(d)));
}

Stream new_thread_local_stream(Device d) {
  return from_mlx(mlx::core::new_thread_local_stream(to_mlx(d)));
}

Stream stream_from_thread_local_stream(Stream s) {
  return from_mlx(mlx::core::stream_from_thread_local_stream(to_mlx_thread_local(s)));
}

void set_default_stream(Stream s) {
  mlx::core::set_default_stream(to_mlx(s));
}

rust::Vec<Stream> get_streams() {
  auto streams = mlx::core::get_streams();
  rust::Vec<Stream> out;
  out.reserve(streams.size());
  for (const auto& s : streams) {
    out.push_back(from_mlx(s));
  }
  return out;
}

void clear_streams() {
  mlx::core::clear_streams();
}

// === Transforms ===

void eval_many(rust::Slice<const MlxArray* const> arrays) {
  std::vector<MlxArray> vec;
  vec.reserve(arrays.size());
  for (size_t i = 0; i < arrays.size(); ++i) {
    vec.push_back(*arrays[i]);  // copy ctor — refcount-shared, cheap
  }
  mlx::core::eval(std::move(vec));
}

void async_eval_many(rust::Slice<const MlxArray* const> arrays) {
  std::vector<MlxArray> vec;
  vec.reserve(arrays.size());
  for (size_t i = 0; i < arrays.size(); ++i) {
    vec.push_back(*arrays[i]);  // copy ctor — refcount-shared, cheap
  }
  mlx::core::async_eval(std::move(vec));
}

void synchronize() {
  mlx::core::synchronize();
}

void synchronize_stream(Stream s) {
  mlx::core::synchronize(to_mlx(s));
}

void synchronize_thread_local_stream(Stream s) {
  mlx::core::synchronize(to_mlx_thread_local(s));
}

void clear_cache() {
  mlx::core::clear_cache();
}

}  // namespace cxx_mlx
